### PR TITLE
ci(docker): rework workflow (3rd time)

### PR DESCRIPTION
**Pull request by hero-intelligent**
### Background

Thanks for your patience. Maybe this will help. Only the tested will be squashed into one commit and pushed here. I'll try not to close pr so frequently. Feel free to review my working status in my repo. 

### Checklist

- [X] The Pull Request has been fully tested
- [ ] There's an entry in the CHANGELOGS
- [ ] There is a user-facing docs PR against https://github.com/daeuniverse/daed

### Full Changelogs

- ci(docker): docker build once and push to multiple repos
- ci(docker): docker build and push with tags
- ci(docker): come back arm64 platform support (Experimental)
- ci(docker): publish dockerhub image
- ci(docker): action version upgrade
- ci(docker): multiple action trigger
- ci(docker): build with tag on dashboard
- ci(docker): i386, armv7 armv8 platform support (Experimental)

### Issue Reference

https://github.com/daeuniverse/daed/issues/280
https://github.com/daeuniverse/daed/pull/267
https://github.com/daeuniverse/daed/pull/286

### Test Result

[Action page](https://github.com/hero-intelligent/daed/actions/runs/6247355789/job/16959768528)
[Dockerhub page](https://hub.docker.com/r/huangqizigailee/daed-test-test/tags)
[Download log archive](https://github.com/hero-intelligent/daed/suites/16364107338/logs?attempt=1)
[View raw logs](https://github.com/hero-intelligent/daed/commit/37a2ba27e8ff6601f36314fb90781399332d1f2e/checks/16959768528/logs)

It should work, but we need cooperation.

### CODE REVIEW

**Comment by yqlbu** [comment]
@hero-intelligent When it comes to `build candidates`, I think we ONLY need to take the following architectures into consideration:

- amd64
- arm64

File: publish.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,47 @@
+
+# ATTENTION This part below is for publishing purpose only
+# NOTE in the as parameter in FROM,
+# the `--` between `web` and `has`, and `dokerfile` in which letter c is missing
+# I do it on purpose so that it is not convenient for self build to avoid unnecessary problem
+
+FROM golang:1.21-bookworm as build-bundle-directly-for-web--has-been-built-from-outside-of-dokerfile
```
Comment (yqlbu) [inline publish.Dockerfile:7]:
```suggestion
FROM golang:1.21-bookworm as base
```

File: publish.Dockerfile [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+# build bundle process
+ENV CGO_ENABLED=0
+ENV CLANG=clang-15
+ARG DAED_VERSION=self-build
+
+WORKDIR /build
+
+COPY ./dist/ ./web/
+COPY ./wing/ ./wing/
+# RUN git clone --recursive https://github.com/daeuniverse/dae-wing.git wing
```
Comment (yqlbu) [inline publish.Dockerfile:22]:
```suggestion
```

File: publish.Dockerfile [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+RUN make APPNAME=daed VERSION=$DAED_VERSION OUTPUT=daed WEB_DIST=/build/web/ bundle
+
+
+
+# ATTENTION This part below is for publishing purpose only
+# NOTE in the from parameter in COPY,
+# the `--` between `web` and `has`, and `dokerfile` in which letter c is missing
+# I do it on purpose so that it is not convenient for self build to avoid unnecessary problem
+
+FROM alpine as this-image-is-built-for-final-publish-with-support-of-multiple-platforms
```
Comment (yqlbu) [inline publish.Dockerfile:35]:
```suggestion
FROM alpine as prod
```

File: publish.Dockerfile [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+FROM golang:1.21-bookworm as build-bundle-directly-for-web--has-been-built-from-outside-of-dokerfile
+
+RUN \
+    apt-get update; apt-get install -y git make llvm-15 clang-15; \
+    apt-get clean autoclean && apt-get autoremove -y && rm -rf /var/lib/{apt,dpkg,cache,log}/
+
+# build bundle process
+ENV CGO_ENABLED=0
+ENV CLANG=clang-15
+ARG DAED_VERSION=self-build
```
Comment (yqlbu) [inline publish.Dockerfile:16]:
```suggestion
ARG DAED_VERSION
```

File: publish.Dockerfile [hunk 5]
```diff
@@ -0,0 +1,47 @@
+
```
Comment (yqlbu) [inline publish.Dockerfile:1]:
```suggestion
ARG DAED_VERSION
```

We pass the version from CI as `build-arg`

**Review by yqlbu — CHANGES_REQUESTED** [review]

**Comment by yqlbu** [comment]
From your sample workflow run, everything meets our expectations except the list of build candidates (platforms). Ideally, we should ONLY consider offering this containerization deployment alternative to `amd64` and `arm64` users and ignoring all other unnecessary platforms given the consideration of less common adoption of other platforms such as `ppc64le`.

File: .github/workflows/publish-docker-image.yml [hunk 6]
```diff
@@ -2,61 +2,151 @@ name: Publish Docker Image
...
   workflow_dispatch:
+  push:
+    branches:
+      - "main"
   release:
-    types: [prereleased]
+    types: [prereleased,released]
+
+env:
+  REGISTRY_IMAGE: daeuniverse/daed
```
Comment (kunish) [inline .github/workflows/publish-docker-image.yml:12]: This variable is no longer necessary

File: .github/workflows/publish-docker-image.yml [hunk 7]
```diff
@@ -2,61 +2,151 @@ name: Publish Docker Image
...
+
 
 jobs:
+
+  build-web:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+        # with:
+        #   submodules: 'recursive'
```
Comment (kunish) [inline .github/workflows/publish-docker-image.yml:22]: We did it on purpose, as those source code files under submodules are also used by linux distro packages, such as archlinux and gentoo. For example, the packaging process of gentoo is done inside a isolated environment, which means we can't do git clone as we desire.

**Review by kunish — CHANGES_REQUESTED** [review]

File: .github/workflows/publish-docker-image.yml (hunk 7 above)
Comment (kunish) [reply .github/workflows/publish-docker-image.yml:22]: Yeah, they are requested by package maintainers

**Review by kunish — APPROVED** [review]

**Comment by kunish** [comment]
Awesome, let's ship it!

**Comment by kunish** [comment]
> This build process is time consuming, which takes about one hour. You can disable `on push` after this run time if you don't need to build after every single update.

OK

**Comment by yqlbu** [comment]
@hero-intelligent I will update the workflow definition accordingly. Stay tuned.

**Pull request merged**